### ISSUE TITLE
Automatically detect different legacy file formats

### ISSUE THREAD

**Issue by TomNicholas** [zarr-developers/VirtualiZarr#10]
We should design the backends so that the user doesn't have to specifically state which type of file they are trying to open. I believe pangeo-forge has some logic for detecting the file format automatically.

Original kerchunk gripe here: https://github.com/fsspec/kerchunk/issues/376

**Comment by TomNicholas** [comment]
This was partially closed by #43, which can auto-detect the difference between netCDF3 and netCDF4 at least.

**Comment by TomNicholas** [comment]
Copying [Ryan's suggestion](https://github.com/TomNicholas/VirtualiZarr/pull/43#discussion_r1543415167) here so we don't forget to try it:

You can do this without opening the file with netCDF at all. Both file types have a "magic" at the beginning of the file.

```python
def guess_file_type(fp) -> FileType:
    magic = fp.read(4)
    fp.seek(0)
    if magic[:3] == b"CDF":
        return FileType.netcdf3
    elif magic == b"\x89HDF":
        return FileType.hdf5
    else:
        raise ValueError(f"Unknown file type - magic {magic}")
```

**Comment by TomNicholas** [comment]
I think this was closed by #143
